vnext/ReactUWP/Modules/AppearanceModule.cpp

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "AppearanceModule.h"
+
+#include <winrt/Windows.UI.ViewManagement.h>
+
+using Application = winrt::Windows::UI::Xaml::Application;
+using ApplicationTheme = winrt::Windows::UI::Xaml::ApplicationTheme;
+using UISettings = winrt::Windows::UI::ViewManagement::UISettings;
+
+using Method = facebook::xplat::module::CxxModule::Method;
+
+namespace react::uwp {
+
+AppearanceChangeListener::AppearanceChangeListener(std::weak_ptr<IReactInstance> &&reactInstance) noexcept
+    : Mso::ActiveObject<>(Mso::DispatchQueue::MainUIQueue()), m_weakReactInstance(std::move(reactInstance)) {}
+
+void AppearanceChangeListener::Initialize() noexcept {
+  m_currentTheme = Application::Current().RequestedTheme();
+
+  // UISettings will notify us on a background thread regardless of where we construct it or register for events. Let it
+  // be constructed before init, but redirect callbacks to the UI thread where we can check app theme.
+  m_uiSettings.ColorValuesChanged([weakThis{Mso::WeakPtr(this)}](const auto &, const auto &) noexcept {
+    if (auto strongThis = weakThis.GetStrongPtr()) {
+      strongThis->InvokeInQueueStrong([strongThis]() noexcept { strongThis->OnColorValuesChanged(); });
+    }
+  });
+
+  m_inited.Set();
+}
+
+const char *AppearanceChangeListener::GetColorScheme() noexcept {
+  // Wait until we have an initial valid value from the UI thread before returning anything
+  m_inited.Wait();
```

Comment:
>.Wait(); [](start = 10, length = 8)

It is ugly. It would be better not to lock the JS thread. It could be a serious perf bottleneck.
A better approach to initialize it upfront as we do with other UI-aware native modules.

Reply:
The problem is that we initialize the native modules on-demand and this code is always going to cause per overhead by stopping JS thread.
Please see how I solve a similar issue with the Linking module: LinkingManagerModule.cpp
There is a static part that always has information available and to provide notifications. The native modules picks it up and subscribes to events without waiting.

---
In reply to: [409881033](https://github.com/microsoft/react-native-windows/pull/4622#discussion_r409881033) [](ancestors = 409881033)

Reply:
I went back and forward on this, I don't think this is ever a perf issue because the UI code to initialize the listener will fire long before the theme is requested, so the lock is just a safety measure. In general I agree that I don't like locking though.

FWIW, any synchronous native module call will already block the JS thread until the native thread can pump all messages and respond. That's not strictly the same as blocking on the UI thread admittedly.

The advantage we get here is that the threadiness of this module becomes an implementation detail that consumers don't need to worry about. E.g. we won't need to do any special initialization shenanigans when refactoring all of this later.

Reply:
I thought about this more overnight. I still don't think it's likely we will ever lock, but at the point where it's not trivial to reason about, we probably shouldn't be using locks. I will change this to follow the pattern of other modules. I'm interested to see the end what we end up coming up with for this problem in native modules 2.0.